### PR TITLE
Extend range of communicator IDs for the RDMA protocol

### PR DESCRIPTION
**Pull request by AmedeoSapio**
*Issue #*

The RDMA protocol is currently using 12 bits for the communicator ID, limiting the number of communicators to 4K, which can be hit in large deployments.

*Description of changes:*

This PR is hanging the number of bits used for the communicator ID from 12 to 18 to reduce
the chance of running out of comm IDs. To fit the comm ID in the immediate data,
the message sequence number has been reduced from 16 bits to 10, which is still
more than enough since usually the number of inflight messages is not that high.
The msg_seq_num space can be further reduced if needed.

With the 16 bits msg_seq_num the msgbuff relied on the wraparound of 16 bits integers,
so to reduce the number of bits of msg_seq_num we need to make the wrapping around
explicit and implement a subtraction to deal with the wraparound case. This makes
the msgbuff more generic, as it can now support any msg_seq_num space and any
number of inflight messages (not only powers of 2).

The msgbuff unit test has also been extended to test the case when message sequence numbers wrap around.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/nccl_ofi_msgbuff.h [hunk 1]
```diff
@@ -74,7 +74,9 @@ typedef struct {
 	// Element storage buffer. Allocated in msgbuff_init
 	nccl_ofi_msgbuff_elem_t *buff;
 	// Number of elements in storage buffer
-	uint16_t buff_size;
+	uint16_t max_inprogress;
```
Comment (bwbarrett) [inline include/nccl_ofi_msgbuff.h:77]: I do not get "maximum size of the storage buffer" from `max_inprogress`.  I think this needs a better name.

File: include/nccl_ofi_msgbuff.h [hunk 2]
```diff
@@ -74,7 +74,9 @@ typedef struct {
...
+	// Maximum number of elements tracked by this msgbuff
+	uint16_t size;
```
Comment (bwbarrett) [inline include/nccl_ofi_msgbuff.h:79]: This needs a better comment.

File: include/nccl_ofi_msgbuff.h [hunk 3]
```diff
@@ -84,11 +86,13 @@ typedef struct {
...
 /**
- * Allocates and initializes a new message buffer. Buffer size should be a power of 2.
+ * Allocates and initializes a new message buffer.
+ * @param max_inprogress number of elements in storage buffer
+ * @param size number of elements tracked by this msgbuff
  *
  * @return a new msgbuff, or NULL if initialization failed
  */
-nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t buffer_size);
+nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t max_inprogress, uint16_t size);
```
Comment (bwbarrett) [inline include/nccl_ofi_msgbuff.h:95]: From reading the documentation for this function, it's not really clear why we need to set both max_inprogress and size.  It's even less clear what these actually mean.  I think the docs need some behavior description.  And probably different names (see above comment).

File: src/nccl_ofi_msgbuff.c [hunk 4]
```diff
@@ -41,12 +46,28 @@ nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t buffer_size)
...
+static inline void increase_with_wraparound(const nccl_ofi_msgbuff_t *msgbuff, uint16_t *ctr)
+{
+	if (OFI_UNLIKELY(*ctr == (msgbuff->size - 1))) {
+		*ctr = 0;
+	} else {
+		++(*ctr);
+	}
+}
+
+static inline uint16_t subtract(const nccl_ofi_msgbuff_t *msgbuff, const uint16_t minuend, const uint16_t subtrahend)
```
Comment (bwbarrett) [inline src/nccl_ofi_msgbuff.c:66]: This function needs a different name.  This is not a simple subtraction.

File: src/nccl_ofi_msgbuff.c [hunk 5]
```diff
@@ -41,12 +46,28 @@ nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t buffer_size)
...
+			free(msgbuff->buff);
+		}
 		free(msgbuff);
 	}
 	return NULL;
 }
 
+static inline void increase_with_wraparound(const nccl_ofi_msgbuff_t *msgbuff, uint16_t *ctr)
+{
+	if (OFI_UNLIKELY(*ctr == (msgbuff->size - 1))) {
```
Comment (bwbarrett) [inline src/nccl_ofi_msgbuff.c:59]: This seems overly complicated.  Why are we not just forcing size to be a power of two and doing masked math like we used to?

File: src/nccl_ofi_rdma.c [hunk 6]
```diff
@@ -37,7 +37,7 @@ static pthread_mutex_t topo_file_lock = PTHREAD_MUTEX_INITIALIZER;
 /*
  * @brief	Number of bits used for the communicator ID
  */
-#define NUM_COMM_ID_BITS           ((uint64_t)12)
+#define NUM_COMM_ID_BITS           ((uint64_t)18)
```
Comment (bwbarrett) [inline src/nccl_ofi_rdma.c:40]: the msgbuf changes and the comm id size changes should be different commits.

File: tests/unit/msgbuff.c [hunk 7]
```diff
@@ -9,12 +9,28 @@
 #include "test-common.h"
 #include "nccl_ofi_msgbuff.h"
 
+static inline uint16_t subtract(const uint16_t size, const uint16_t minuend, const uint16_t subtrahend)
```
Comment (bwbarrett) [inline tests/unit/msgbuff.c:12]: The msgbuff unit test commit should be part of the commit that changes the msgbuff interface.

Also, the subtract function irritates me for the same reason called out in the previous commit, and let's figure out how not to have it exist in two places.

**Review by bwbarrett — CHANGES_REQUESTED** [review]
Note that I only reviewed the last two commits, since the other ones are covered in a different PR.  My assumption is that this PR will rebase once that one is committed.

File: src/nccl_ofi_msgbuff.c [hunk 8]
```diff
@@ -119,7 +140,7 @@ nccl_ofi_msgbuff_result_t nccl_ofi_msgbuff_insert(nccl_ofi_msgbuff_t *msgbuff,
 		uint16_t msg_index, void *elem, nccl_ofi_msgbuff_elemtype_t type,
 		nccl_ofi_msgbuff_status_t *msg_idx_status)
 {
-	if (!msgbuff) {
+	if (OFI_UNLIKELY(!msgbuff)) {
```
Comment (rauteric) [inline src/nccl_ofi_msgbuff.c:143]: Yeah, or maybe even just an assert here? Same below.

File: src/nccl_ofi_msgbuff.c [hunk 9]
```diff
@@ -41,12 +46,28 @@ nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t buffer_size)
...
-		if (msgbuff->buff) free(msgbuff->buff);
+		if (msgbuff->buff) {
+			free(msgbuff->buff);
+		}
 		free(msgbuff);
 	}
 	return NULL;
 }
 
+static inline void increase_with_wraparound(const nccl_ofi_msgbuff_t *msgbuff, uint16_t *ctr)
```
Comment (rauteric) [inline src/nccl_ofi_msgbuff.c:57]: `increment_with_wraparound`?

File: src/nccl_ofi_rdma.c [hunk 10]
```diff
@@ -3070,7 +3073,7 @@ static int recv(nccl_net_ofi_recv_comm_t *recv_comm, int n, void **buffers,
 	/* Return request to NCCL */
 	*base_req = (nccl_net_ofi_req_t *)req;
 	/* Increment next_msg_seq_num for next call */
-	++(r_comm->next_msg_seq_num);
+	r_comm->next_msg_seq_num = (r_comm->next_msg_seq_num + 1) % MSG_SEQ_NUM_SIZE;
```
Comment (rauteric) [inline src/nccl_ofi_rdma.c:3076]: Since `MSG_SEQ_NUM_SIZE` is a power of two, would it be more efficient to do this with a bitmask? Same below.

File: include/nccl_ofi_msgbuff.h (hunk 3 above)
Comment (AmedeoSapio) [reply include/nccl_ofi_msgbuff.h]: I improved the documentation to make it clearer what this is doing. Let me know if it still isn't clear enough.

File: src/nccl_ofi_rdma.c [hunk 11]
```diff
@@ -51,13 +51,13 @@ static pthread_mutex_t topo_file_lock = PTHREAD_MUTEX_INITIALIZER;
...
  *
- * | 4-bit segment count | 12-bit comm ID | 16-bit msg_seq_num |
+ * | 6-bit unused | 4-bit segment count | 12-bit comm ID | 10-bit msg_seq_num |
  *
  * - Segment count: number of RDMA writes that will be delivered as part of this message
  * - Comm ID: the ID for this communicator
  * - Message sequence number: message identifier
  */
-#define NUM_MSG_SEQ_NUM_BITS ((uint64_t) 16)
+#define NUM_MSG_SEQ_NUM_BITS       ((uint64_t)10)
```
Comment (bwbarrett) [inline src/nccl_ofi_rdma.c:60]: why include this part in this patch?  Just leave it at 16 adjust the bit fields in this file in the next commit.  Make this commit *just* about changing the msgbuf interface.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: src/nccl_ofi_msgbuff.c [hunk 12]
```diff
@@ -5,32 +5,38 @@
...
+nccl_ofi_msgbuff_t *nccl_ofi_msgbuff_init(uint16_t max_inprogress, uint16_t bit_width)
 {
 	nccl_ofi_msgbuff_t *msgbuff = NULL;
 
-	if (buffer_size == 0) {
-		NCCL_OFI_WARN("Refusing to allocate empty buffer");
-		goto error;
+	if (max_inprogress == 0 || (uint16_t)(1 << bit_width) <= 2 * max_inprogress) {
+		NCCL_OFI_WARN("Wrong parameters for msgbuff_init max_inprogress %" PRIu16 " bit_width %" PRIu16 "",
+			      max_inprogress, bit_width);
```
Comment (rauteric) [inline src/nccl_ofi_msgbuff.c:20]: Missing `goto error`?

Comment (AmedeoSapio) [reply src/nccl_ofi_msgbuff.c:20]: Good catch! Fixed it

**Review by rauteric — APPROVED** [review]

**Review by bwbarrett — APPROVED** [review]
